<a href="https://colab.research.google.com/github/quinnie-nguyen/Patent-Data-and-Social-Network-Analysis-using-R/blob/master/demo_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This notebook uses cleaned data, i.e, a)drop all features that have more than 80% misssing values; b) drop rows that have more than 50% features missing; c)imputation by mean, mode, median.


Let's see how it works!


!!!!!!!!! minimum accuracy must be **78.3%**

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
# import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.ensemble import VotingClassifier
from sklearn.pipeline import Pipeline
from functools import partial
from collections import Counter
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.tree import DecisionTreeClassifier
from imblearn.over_sampling import SMOTE
from imblearn.over_sampling import BorderlineSMOTE
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import StratifiedKFold
from imblearn.over_sampling import RandomOverSampler
from imblearn.under_sampling import RandomUnderSampler

In [ ]:
# load data
df = pd.read_csv('/content/gdrive/My Drive/WIDS/cleaned_data_0702.csv')
dictionary = pd.read_excel('/content/gdrive/My Drive/WIDS/DataDictionaryWiDS2021-fixed-thu-tu.xlsx')

In [ ]:
dictionary

class                    feature  ...    example rank
0   APACHE covariate             glucose_apache  ...          5    1
1               labs             h1_glucose_max  ...          5    2
2               labs             h1_glucose_min  ...          5    2
3               labs             d1_glucose_max  ...          5    3
4               labs             d1_glucose_min  ...          5    3
5        demographic                     gender  ...          F    4
6        demographic                     height  ...        180    5
7        demographic                     weight  ...         80    5
8        demographic                  ethnicity  ...  Caucasian    6
9             vitals     d1_diasbp_invasive_max  ...         60    7
10            vitals     d1_diasbp_invasive_min  ...         60    7
11            vitals              d1_diasbp_max  ...         60    8
12            vitals              d1_diasbp_min  ...         60    8
13            vitals  d1_diasbp_noninvasive_max  ...         60    9
14            vitals  d1_diasbp_noninvasive_min  ...         60    9
15            vitals      d1_sysbp_invasive_max  ...        120   10
16            vitals      d1_sysbp_invasive_min  ...        120   10
17            vitals               d1_sysbp_max  ...        120   11
18            vitals               d1_sysbp_min  ...        120   11
19            vitals   d1_sysbp_noninvasive_max  ...        120   12
20            vitals   d1_sysbp_noninvasive_min  ...        120   12
21  APACHE covariate          heart_rate_apache  ...         75   13
22            vitals           d1_heartrate_max  ...         75   14
23            vitals           d1_heartrate_min  ...         75   14
24            vitals     h1_diasbp_invasive_max  ...         60   15
25            vitals     h1_diasbp_invasive_min  ...         60   15
26            vitals              h1_diasbp_max  ...         60   16
27            vitals              h1_diasbp_min  ...         60   16
28            vitals  h1_diasbp_noninvasive_max  ...         60   17
29            vitals  h1_diasbp_noninvasive_min  ...         60   17
30            vitals      h1_sysbp_invasive_max  ...        120   18
31            vitals      h1_sysbp_invasive_min  ...        120   18
32            vitals               h1_sysbp_max  ...        120   19
33            vitals               h1_sysbp_min  ...        120   19
34            vitals   h1_sysbp_noninvasive_max  ...        120   20
35            vitals   h1_sysbp_noninvasive_min  ...        120   20
36            vitals           h1_heartrate_max  ...         75   21
37            vitals           h1_heartrate_min  ...         75   21
38              labs             d1_albumin_max  ...         30   22
39              labs             d1_albumin_min  ...         30   22
40              labs             h1_lactate_max  ...          1   24
41              labs             h1_lactate_min  ...          1   24
42              labs             d1_lactate_max  ...          1   25
43              labs             d1_lactate_min  ...          1   25
44              labs           d1_bilirubin_max  ...         20   26
45              labs           d1_bilirubin_min  ...         20   26
46              labs             h1_albumin_max  ...         30   27
47              labs             h1_albumin_min  ...         30   27
48              labs           h1_bilirubin_max  ...         20   28
49              labs           h1_bilirubin_min  ...         20   28

[50 rows x 7 columns]

In [ ]:
con_fea = ['age','bmi','glucose_apache','h1_glucose_max','h1_glucose_min','d1_glucose_max','d1_glucose_min','d1_diasbp_invasive_max','d1_diasbp_invasive_min','d1_diasbp_max',
           'd1_diasbp_min','d1_diasbp_noninvasive_max','d1_diasbp_noninvasive_min','d1_sysbp_invasive_max','d1_sysbp_invasive_min','d1_sysbp_max','d1_sysbp_min',
           'd1_sysbp_noninvasive_max','d1_sysbp_noninvasive_min','heart_rate_apache','d1_heartrate_max','d1_heartrate_min','h1_diasbp_invasive_max','h1_diasbp_invasive_min',
           'h1_diasbp_max','h1_diasbp_min','h1_diasbp_noninvasive_max','h1_diasbp_noninvasive_min','h1_sysbp_invasive_max','h1_sysbp_invasive_min','h1_sysbp_max',
           'h1_sysbp_min','h1_sysbp_noninvasive_max','h1_sysbp_noninvasive_min','h1_heartrate_max','h1_heartrate_min','d1_albumin_max','d1_albumin_min',
           'h1_lactate_max','h1_lactate_min','d1_lactate_max','d1_lactate_min','d1_bilirubin_max','d1_bilirubin_min','h1_albumin_max','h1_albumin_min',
           'h1_bilirubin_max','h1_bilirubin_min','ethnicity_Asian','ethnicity_Caucasian','ethnicity_Hispanic','ethnicity_Native American','ethnicity_Other/Unknown',
           'gender_M']

remaining_fea = df.columns

selected_fea = [col for col in con_fea if col in remaining_fea]
selected_fea

['age',
 'bmi',
 'glucose_apache',
 'h1_glucose_max',
 'h1_glucose_min',
 'd1_glucose_max',
 'd1_glucose_min',
 'd1_diasbp_invasive_max',
 'd1_diasbp_invasive_min',
 'd1_diasbp_max',
 'd1_diasbp_min',
 'd1_diasbp_noninvasive_max',
 'd1_diasbp_noninvasive_min',
 'd1_sysbp_invasive_max',
 'd1_sysbp_invasive_min',
 'd1_sysbp_max',
 'd1_sysbp_min',
 'd1_sysbp_noninvasive_max',
 'd1_sysbp_noninvasive_min',
 'heart_rate_apache',
 'd1_heartrate_max',
 'd1_heartrate_min',
 'h1_diasbp_max',
 'h1_diasbp_min',
 'h1_diasbp_noninvasive_max',
 'h1_diasbp_noninvasive_min',
 'h1_sysbp_max',
 'h1_sysbp_min',
 'h1_sysbp_noninvasive_max',
 'h1_sysbp_noninvasive_min',
 'h1_heartrate_max',
 'h1_heartrate_min',
 'd1_albumin_max',
 'd1_albumin_min',
 'd1_lactate_max',
 'd1_lactate_min',
 'd1_bilirubin_max',
 'd1_bilirubin_min',
 'ethnicity_Asian',
 'ethnicity_Caucasian',
 'ethnicity_Hispanic',
 'ethnicity_Native American',
 'ethnicity_Other/Unknown',
 'gender_M']

In [ ]:
df.head()

Unnamed: 0   age  ...  icu_type_SICU  diabetes_mellitus
0           0  68.0  ...              0                1.0
1           1  77.0  ...              0                1.0
2           2  25.0  ...              0                0.0
3           3  81.0  ...              0                0.0
4           4  67.0  ...              0                1.0

[5 rows x 173 columns]

In [ ]:
 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=1, stratify=y)

# Building models applying Stratified Split!!!

In [ ]:
y = df.iloc[:,-1]

In [ ]:
X = df.loc[:, selected_fea]

In [ ]:
y.head()

0    1.0
1    1.0
2    0.0
3    0.0
4    1.0
Name: diabetes_mellitus, dtype: float64

In [ ]:
X.head()

glucose_apache  h1_glucose_max  ...  ethnicity_Other/Unknown  gender_M
0       168.00000      164.830653  ...                        0         1
1       145.00000      145.000000  ...                        0         0
2       159.31134      167.721828  ...                        0         0
3       185.00000      167.721828  ...                        0         0
4       156.00000      164.830653  ...                        0         1

[5 rows x 42 columns]

In [ ]:
y.value_counts()

0.0    100699
1.0     27958
Name: diabetes_mellitus, dtype: int64

In [ ]:
# split imbalanced dataset into train and test sets with stratification
from collections import Counter
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
# create dataset
print(Counter(y))
# split into train test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=1, stratify=y)
print(Counter(y_train))
print(Counter(y_test))

Counter({0.0: 100699, 1.0: 27958})
Counter({0.0: 80559, 1.0: 22366})
Counter({0.0: 20140, 1.0: 5592})


In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import StratifiedKFold
from imblearn.over_sampling import RandomOverSampler
from imblearn.under_sampling import RandomUnderSampler


In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
logisticRegr = LogisticRegression(solver = 'newton-cg', max_iter = 2000)
logisticRegr.fit(X_train, y_train)
y_pred = logisticRegr.predict(X_test)
accuracy_score(y_test, y_pred)

0.8049510337323177

In [ ]:
len(X.columns)

42

In [ ]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
accuracy_score(y_test, y_pred)

0.8133452510492771

In [ ]:
from xgboost import XGBClassifier
clf4 = XGBClassifier(max_depth=30, n_estimators=42,  objective='multi:softprob', learning_rate=0.3, num_class=2)
clf4.fit(X_train, y_train)
y_pred = clf4.predict(X_test)
accuracy_score(y_test, y_pred)

0.8070884501787657

In [ ]:
#from sklearn.neighbors import (NeighborhoodComponentsAnalysis, KNeighborsClassifier)
#ca = NeighborhoodComponentsAnalysis(random_state=42)
#nca.fit(X_train, y_train)
#y_nca = nca.predict(X_test)
#accuracy_score(y_test, y_nca)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)
y_knn = knn.predict(X_test)

In [ ]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_knn)

0.7895616353178921

In [ ]:
from sklearn.ensemble import VotingClassifier
from sklearn.pipeline import Pipeline
from functools import partial

models = list()
# create the pipeline
#pipe = Pipeline([('ContinuousFeturesSelector',ce), ('m', clf3)])
# add as a tuple to the list of models for voting
models.append(('rf', clf))
models.append(('knn', knn))
models.append(('lg', logisticRegr))

# define the voting ensemble

model = VotingClassifier(estimators=models, voting='soft', weights=[4,4,3])

model.fit(X_train.values, y_train)
print(f"accuracy score voting is {accuracy_score(y_test, model.predict(X_test.values))}")

accuracy score voting is 0.8096144877972952


In [ ]:
clf = RandomForestClassifier(
    n_estimators=50,
    criterion='gini',
    max_depth=5,
    min_samples_split=2,
    min_samples_leaf=1,
    min_weight_fraction_leaf=0.0,
    max_features='auto',
    max_leaf_nodes=None,
    min_impurity_decrease=0.0,
    min_impurity_split=None,
    bootstrap=True,
    oob_score=False,
    n_jobs=-1,
    random_state=0,
    verbose=0,
    warm_start=False,
    class_weight='balanced'
)

In [ ]:
from sklearn.model_selection import StratifiedKFold
cv = StratifiedKFold(n_splits=5, random_state=123, shuffle=True)
results = pd.DataFrame(columns=['training_score', 'test_score'])
accuracy = []
    
for (train, test), i in zip(cv.split(X, y), range(5)):
    clf.fit(X.iloc[train], y.iloc[train])
    y_pred = clf.predict(X.iloc[test])
    accuracy.append(accuracy_score(y.iloc[test], y_pred))


In [ ]:
accuracy

[0.810624902844707,
 0.8101974195554174,
 0.8126384516730791,
 0.8098791341183786,
 0.813998678636664]

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_knn))

              precision    recall  f1-score   support

         0.0       0.84      0.91      0.87     20140
         1.0       0.52      0.36      0.42      5592

    accuracy                           0.79     25732
   macro avg       0.68      0.63      0.65     25732
weighted avg       0.77      0.79      0.77     25732



# Building models based on resample dataset (balance the data wrt target)

## Oversampling method

In [ ]:
X = df.loc[:, selected_fea]
y = df.iloc[:,-1]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=1, stratify=y)

(128657, 42)

In [ ]:
y.value_counts()/len(y)

0.0    0.782694
1.0    0.217306
Name: diabetes_mellitus, dtype: float64

In [ ]:
y_train.value_counts()/len(y_train)

0.0    0.782696
1.0    0.217304
Name: diabetes_mellitus, dtype: float64

In [ ]:
 pip install imbalanced-learn

In [ ]:
from imblearn.over_sampling import RandomOverSampler
# define oversampling strategy
oversample = RandomOverSampler(sampling_strategy='minority')
# fit and apply the transform
X_over, y_over = oversample.fit_resample(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)


In [ ]:
from collections import Counter
print(Counter(y_over))

Counter({0.0: 80559, 1.0: 80559})


In [ ]:
X_over = pd.DataFrame(X_over)
X_over.columns = X.columns
X_over.head()

age        bmi  ...  ethnicity_Other/Unknown  gender_M
0  78.0  25.448202  ...                      0.0       1.0
1  44.0  33.214352  ...                      0.0       1.0
2  71.0  22.955862  ...                      0.0       1.0
3  46.0  15.937719  ...                      0.0       0.0
4  69.0  22.173655  ...                      0.0       0.0

[5 rows x 44 columns]

In [ ]:
df['diabetes_mellitus'].value_counts()

0.0    100699
1.0     27958
Name: diabetes_mellitus, dtype: int64

In [ ]:
#from sklearn.utils import resample

#tr_majority = df[df['diabetes_mellitus'] == 0]
#tr_minority = df[df['diabetes_mellitus'] == 1]

#tr_minority_upsampled = resample(tr_minority,
                                 replace=True,  # sample with replacement
                                 n_samples=100000,  # to match majority class 84000
                                 random_state=1)  # reproducible results

# Combine majority class with upsampled minority class
#tr_upsampled = pd.concat([tr_majority, tr_minority_upsampled])

# Display new class counts
#tr_upsampled.diabetes_mellitus.value_counts()


In [ ]:
#X_upsampled = tr_upsampled.loc[:,selected_fea]
#y_upsampled = tr_upsampled.iloc[:,-1]
#from sklearn.model_selection import train_test_split
# split into train test sets
#X_train, X_test, y_train, y_test = train_test_split(X_upsampled, y_upsampled, test_size=0.20, random_state=1)

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
logisticRegr = LogisticRegression(solver = 'newton-cg', max_iter = 2000)
logisticRegr.fit(X_over, y_over)
y_pred = logisticRegr.predict(X_test)
accuracy_score(y_test, y_pred)

0.7557515933468055

!!! Probably linear classificator is NOT suitable for our problem 

In [ ]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()
clf.fit(X_over, y_over)
y_pred = clf.predict(X_test)
accuracy_score(y_test, y_pred)

0.8129566298771957

RF

*   oversampling the whole dataset --> 85.6%
*   oversampling just training set --> 81.05%


In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_over, y_over)
y_knn = knn.predict(X_test)
accuracy_score(y_test, y_knn)

0.6828462614643246


*   78.5%
*   68.3%



In [ ]:
from xgboost import XGBClassifier
clf4 = XGBClassifier(max_depth=30, n_estimators=42,  objective='multi:softprob', learning_rate=0.3, num_class=2)
clf4.fit(X_over, y_over)
y_pred = clf4.predict(X_test)
accuracy_score(y_test, y_pred)

0.8039794808021141

XGBC

*   oversampling the whole dataset --> 92.6%
*   oversampling just training set --> 80.4%


In [ ]:
from sklearn.ensemble import VotingClassifier
from sklearn.pipeline import Pipeline
from functools import partial
from sklearn.metrics import classification_report


models = list()
# create the pipeline
#pipe = Pipeline([('ContinuousFeturesSelector',ce), ('m', clf3)])
# add as a tuple to the list of models for voting
models.append(('rf', clf))
models.append(('knn', knn))
models.append(('XGBC', clf4))
models.append(('lg', logisticRegr))

# define the voting ensemble

model = VotingClassifier(estimators=models, voting='soft', weights=[3,2,3,2])

model.fit(X_over.values, y_over)
print(f"accuracy score voting is {accuracy_score(y_test, model.predict(X_test.values))}")
print(classification_report(y_test, model.predict(X_test.values)))

accuracy score voting is 0.8031633763407431
              precision    recall  f1-score   support

         0.0       0.87      0.88      0.88     20140
         1.0       0.55      0.52      0.53      5592

    accuracy                           0.80     25732
   macro avg       0.71      0.70      0.70     25732
weighted avg       0.80      0.80      0.80     25732



VC


*   Oversampling the whole : 92.22%


##Undersampling method

In [ ]:
from imblearn.under_sampling import RandomUnderSampler
undersample = RandomUnderSampler(sampling_strategy='majority')
# fit and apply the transform
X_under, y_under = undersample.fit_resample(X, y)

/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)


In [ ]:
X_under = pd.DataFrame(X_under)
X_under.columns = X_train.columns

In [ ]:
print(len(y_under))
print(len(y))
print(Counter(y_under))

55916
128657
Counter({0.0: 27958, 1.0: 27958})


In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
logisticRegr = LogisticRegression(solver = 'newton-cg', max_iter = 2000)
logisticRegr.fit(X_under, y_under)
y_log = logisticRegr.predict(X_test)
accuracy_score(y_test, y_log)

0.7545857298305612

In [ ]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()
clf.fit(X_under, y_under)
y_rf = clf.predict(X_test)
accuracy_score(y_test, y_rf)

0.8546556816415358

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_under, y_under)
y_knn = knn.predict(X_test)
accuracy_score(y_test, y_knn)

0.7465801336856832

In [ ]:
from xgboost import XGBClassifier
clf4 = XGBClassifier(max_depth=30, n_estimators=42,  objective='multi:softprob', learning_rate=0.3, num_class=2)
clf4.fit(X_under, y_under)
y_xgbc = clf4.predict(X_test)
accuracy_score(y_test, y_xgbc)

0.8447069796362506

84.62 without age and bmi

In [ ]:
from sklearn.ensemble import VotingClassifier
from sklearn.pipeline import Pipeline
from functools import partial
from sklearn.metrics import classification_report


models = list()
# create the pipeline
#pipe = Pipeline([('ContinuousFeturesSelector',ce), ('m', clf3)])
# add as a tuple to the list of models for voting
models.append(('rf', clf))
models.append(('knn', knn))
models.append(('XGBC', clf4))
models.append(('lg', logisticRegr))

# define the voting ensemble

model = VotingClassifier(estimators=models, voting='soft', weights=[4,1,4,1])

model.fit(X_under.values, y_under)
print(f"accuracy score voting is {accuracy_score(y_test, model.predict(X_test.values))}")
print(classification_report(y_test, model.predict(X_test.values)))

accuracy score voting is 0.8539561635317892
              precision    recall  f1-score   support

         0.0       1.00      0.81      0.90     20140
         1.0       0.60      1.00      0.75      5592

    accuracy                           0.85     25732
   macro avg       0.80      0.91      0.82     25732
weighted avg       0.91      0.85      0.86     25732



##Combining Random Oversampling and Undersampling

In [ ]:
y_train.value_counts()

0.0    80559
1.0    22366
Name: diabetes_mellitus, dtype: int64

In [ ]:
over = RandomOverSampler(sampling_strategy=0.4)
X_com, y_com = over.fit_resample(X_train, y_train)
print(Counter(y_com))
under = RandomUnderSampler(sampling_strategy=0.6)
X_com, y_com = under.fit_resample(X_com, y_com)
print(Counter(y_com))

Counter({0.0: 80559, 1.0: 32223})
Counter({0.0: 53705, 1.0: 32223})


/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)


In [ ]:
X_com = pd.DataFrame(X_com)
X_com.columns = X_train.columns
X_com.head()

age        bmi  ...  ethnicity_Other/Unknown  gender_M
0  66.0  20.414982  ...                      0.0       1.0
1  57.0  41.470491  ...                      0.0       1.0
2  81.0  21.645014  ...                      0.0       1.0
3  71.0  29.923671  ...                      0.0       0.0
4  87.0  23.502142  ...                      0.0       0.0

[5 rows x 44 columns]

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
logisticRegr = LogisticRegression(solver = 'newton-cg', max_iter = 2000)
logisticRegr.fit(X_com, y_com)
y_log = logisticRegr.predict(X_test)
accuracy_score(y_test, y_log)

0.8007927871910462

LG:

*   80.32 (without age and bmi)
*   List item



In [ ]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()
clf.fit(X_com, y_com)
y_rf = clf.predict(X_test)
accuracy_score(y_test, y_rf)

0.8099253847349603

RF


*   80.59
*   List item



In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_com, y_com)
y_knn = knn.predict(X_test)
accuracy_score(y_test, y_knn)

0.7441706824187782

In [ ]:
from xgboost import XGBClassifier
clf4 = XGBClassifier(max_depth=30, n_estimators=42,  objective='multi:softprob', learning_rate=0.3, num_class=2)
clf4.fit(X_com, y_com)
y_xgbc = clf4.predict(X_test)
accuracy_score(y_test, y_xgbc)

0.8000932690812995

XGB 

*   79.737
*   List item



In [ ]:
from sklearn.ensemble import VotingClassifier
from sklearn.pipeline import Pipeline
from functools import partial
from sklearn.metrics import classification_report


models = list()
# create the pipeline
#pipe = Pipeline([('ContinuousFeturesSelector',ce), ('m', clf3)])
# add as a tuple to the list of models for voting
models.append(('rf', clf))
models.append(('knn', knn))
models.append(('XGBC', clf4))
models.append(('lg', logisticRegr))

# define the voting ensemble

model = VotingClassifier(estimators=models, voting='soft', weights=[4,1,2,3])

model.fit(X_com.values, y_com)
print(f"accuracy score voting is {accuracy_score(y_test, model.predict(X_test.values))}")
print(classification_report(y_test, model.predict(X_test.values)))

accuracy score voting is 0.8096533499145033
              precision    recall  f1-score   support

         0.0       0.88      0.88      0.88     20140
         1.0       0.56      0.56      0.56      5592

    accuracy                           0.81     25732
   macro avg       0.72      0.72      0.72     25732
weighted avg       0.81      0.81      0.81     25732



VC


*   80.724
*   Precision (0.56 and 0.87)



In [ ]:
from sklearn.naive_bayes import GaussianNB
GNB = GaussianNB()
GNB.fit(X_com, y_com)
y_gnb = GNB.predict(X_test)
print(accuracy_score(y_test, y_gnb))
print(classification_report(y_test, y_gnb))

0.7868412871133219
              precision    recall  f1-score   support

         0.0       0.85      0.88      0.87     20140
         1.0       0.51      0.44      0.47      5592

    accuracy                           0.79     25732
   macro avg       0.68      0.66      0.67     25732
weighted avg       0.78      0.79      0.78     25732



In [ ]:
from sklearn.naive_bayes import GaussianNB
GNB = GaussianNB()
GNB.fit(X_over, y_over)
y_gnb = GNB.predict(X_test)
print(accuracy_score(y_test, y_gnb))
print(classification_report(y_test, y_gnb))

0.7796906575470232
              precision    recall  f1-score   support

         0.0       0.86      0.86      0.86     20140
         1.0       0.49      0.49      0.49      5592

    accuracy                           0.78     25732
   macro avg       0.68      0.67      0.68     25732
weighted avg       0.78      0.78      0.78     25732



In [ ]:
from sklearn.naive_bayes import GaussianNB
GNB = GaussianNB()
GNB.fit(X_under, y_under)
y_gnb = GNB.predict(X_test)
print(accuracy_score(y_test, y_gnb))
print(classification_report(y_test, y_gnb))

0.7731229597388466
              precision    recall  f1-score   support

         0.0       0.86      0.85      0.85     20140
         1.0       0.48      0.49      0.49      5592

    accuracy                           0.77     25732
   macro avg       0.67      0.67      0.67     25732
weighted avg       0.78      0.77      0.77     25732



In [ ]:
from sklearn import svm
svc = svm.SVC()
svc.fit(X_under, y_under)
y_svc = svc.predict(X_test)
print(accuracy_score(y_test, y_svc))
print(classification_report(y_test, y_svc))

0.7452976838178144
              precision    recall  f1-score   support

         0.0       0.91      0.75      0.82     20140
         1.0       0.45      0.74      0.56      5592

    accuracy                           0.75     25732
   macro avg       0.68      0.74      0.69     25732
weighted avg       0.81      0.75      0.76     25732



### SMOTE Oversampling method

In [ ]:
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.tree import DecisionTreeClassifier
from imblearn.over_sampling import SMOTE
from imblearn.over_sampling import BorderlineSMOTE


In [ ]:
from numpy import mean
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.tree import DecisionTreeClassifier
from imblearn.pipeline import Pipeline
from imblearn.over_sampling import SMOTE

# define pipeline
steps = [('over', SMOTE()), ('model', DecisionTreeClassifier())]
pipeline = Pipeline(steps=steps)
# evaluate pipeline
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
scores = cross_val_score(pipeline, X, y, scoring='roc_auc', cv=cv, n_jobs=-1)
print('Mean ROC AUC: %.3f' % mean(scores))

/usr/local/lib/python3.6/dist-packages/sklearn/externals/six.py:31: FutureWarning: The module is deprecated in version 0.21 and will be removed in version 0.23 since we've dropped support for Python 2.7. Please rely on the official version of six (https://pypi.org/project/six/).
  "(https://pypi.org/project/six/).", FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.neighbors.base module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.neighbors. Anything that cannot be imported from sklearn.neighbors is now part of the private API.
  warnings.warn(message, FutureWarning)


Mean ROC AUC: 0.631


In [ ]:
from imblearn.over_sampling import BorderlineSMOTE
oversample = BorderlineSMOTE()
X_bsmote, y_bsmote = oversample.fit_resample(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)


In [ ]:
X_bsmote = pd.DataFrame(X_bsmote)
X_bsmote.columns = X_train.columns
X_bsmote.shape

(161118, 44)

In [ ]:
from sklearn.naive_bayes import GaussianNB
GNB = GaussianNB()
GNB.fit(X_bsmote, y_bsmote)
y_gnb = GNB.predict(X_test)
print(accuracy_score(y_test, y_gnb))
print(classification_report(y_test, y_gnb))

0.6968754857764651
              precision    recall  f1-score   support

         0.0       0.88      0.71      0.79     20140
         1.0       0.38      0.65      0.48      5592

    accuracy                           0.70     25732
   macro avg       0.63      0.68      0.63     25732
weighted avg       0.77      0.70      0.72     25732



In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_bsmote, y_bsmote)
y_knn = knn.predict(X_test)
print(accuracy_score(y_test, y_knn))
print(classification_report(y_test, y_knn))

0.685916368723768
              precision    recall  f1-score   support

         0.0       0.89      0.68      0.77     20140
         1.0       0.38      0.69      0.49      5592

    accuracy                           0.69     25732
   macro avg       0.63      0.69      0.63     25732
weighted avg       0.78      0.69      0.71     25732



In [ ]:
y_train_knn = knn.predict(X_bsmote)
accuracy_score(y_train_knn, y_bsmote)

In [ ]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()
clf.fit(X_bsmote, y_bsmote)
y_rf = clf.predict(X_test)
accuracy_score(y_test, y_rf)

In [ ]:
from collections import Counter
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler
from imblearn.pipeline import Pipeline
over = SMOTE(sampling_strategy=0.4)
under = RandomUnderSampler(sampling_strategy=0.6)
steps = [('o', over), ('u', under)]
pipeline = Pipeline(steps=steps)
X_scom, y_scom = pipeline.fit_resample(X_train, y_train)
counter = Counter(y_scom)

/usr/local/lib/python3.6/dist-packages/sklearn/externals/six.py:31: FutureWarning: The module is deprecated in version 0.21 and will be removed in version 0.23 since we've dropped support for Python 2.7. Please rely on the official version of six (https://pypi.org/project/six/).
  "(https://pypi.org/project/six/).", FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.neighbors.base module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.neighbors. Anything that cannot be imported from sklearn.neighbors is now part of the private API.
  warnings.warn(message, FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)
/usr/loca

In [ ]:
print(y_train.value_counts())
print(counter)
X_scom = pd.DataFrame(X_scom)
X_scom.columns = X_train.columns

0.0    80559
1.0    22366
Name: diabetes_mellitus, dtype: int64
Counter({0.0: 53705, 1.0: 32223})


In [ ]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()
clf.fit(X_scom, y_scom)
y_rf = clf.predict(X_test)
print(accuracy_score(y_test, y_rf))
print(classification_report(y_test, y_rf))
print(classification_report(y_test, y_rf))

0.8074770713508472

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_scom, y_scom)
y_knn = knn.predict(X_test)
print(accuracy_score(y_test, y_knn))
print(classification_report(y_test, y_knn))

0.7347271879371988
              precision    recall  f1-score   support

         0.0       0.87      0.77      0.82     20140
         1.0       0.42      0.60      0.49      5592

    accuracy                           0.73     25732
   macro avg       0.65      0.69      0.66     25732
weighted avg       0.78      0.73      0.75     25732



In [ ]:
from xgboost import XGBClassifier
clf4 = XGBClassifier(max_depth=30, n_estimators=42,  objective='multi:softprob', learning_rate=0.3, num_class=2)
clf4.fit(X_scom, y_scom)
y_xgbc = clf4.predict(X_test)
print(accuracy_score(y_test, y_xgbc))
print(classification_report(y_test, y_xgbc))

0.7990828540338878
              precision    recall  f1-score   support

         0.0       0.88      0.86      0.87     20140
         1.0       0.54      0.56      0.55      5592

    accuracy                           0.80     25732
   macro avg       0.71      0.71      0.71     25732
weighted avg       0.80      0.80      0.80     25732



In [ ]:
from sklearn.ensemble import VotingClassifier
from sklearn.pipeline import Pipeline
from functools import partial
from sklearn.metrics import classification_report


models = list()
# create the pipeline
#pipe = Pipeline([('ContinuousFeturesSelector',ce), ('m', clf3)])
# add as a tuple to the list of models for voting
models.append(('rf', clf))
models.append(('knn', knn))
models.append(('XGBC', clf4))


# define the voting ensemble

model = VotingClassifier(estimators=models, voting='soft', weights=[4,2,4])

model.fit(X_scom.values, y_scom)
print(f"accuracy score voting is {accuracy_score(y_test, model.predict(X_test.values))}")
print(classification_report(y_test, model.predict(X_test.values)))

accuracy score voting is 0.8040960671537385
              precision    recall  f1-score   support

         0.0       0.88      0.87      0.87     20140
         1.0       0.55      0.57      0.56      5592

    accuracy                           0.80     25732
   macro avg       0.71      0.72      0.72     25732
weighted avg       0.81      0.80      0.81     25732

